lib/permission-api.js

```diff
@@ -0,0 +1,270 @@
+  {
+    name: 'Permission',
+    properties: {
+        id:            { type: 'string' },
+        updatedAt:     { type: 'date' },
+        userId:        { type: 'string' },
+        path:          { type: 'string' },
+        mayRead:       { type: 'bool' },
+        mayWrite:      { type: 'bool' },
+        mayManage:     { type: 'bool' },
+    }
+  }
+];
+
+// Symbols are not supported on RN yet, so we use this for now:
+const specialPurposeRealmsKey = '_specialPurposeRealms';
+
+function getSpecialPurposeRealm(user, realmName, schema) {
+  if (!user.hasOwnProperty(specialPurposeRealmsKey)) {
+    user[specialPurposeRealmsKey] = {};
+  }
+
+  if (user[specialPurposeRealmsKey].hasOwnProperty(realmName)) {
+    return Promise.resolve(user[specialPurposeRealmsKey][realmName]);
+  }
+
+  const url = url_parse(user.server);
+  if (url.protocol === 'http:') {
+    url.set('protocol', 'realm:');
+  } else if (url.protocol === 'https:') {
+    url.set('protocol', 'realms:');
+  } else {
+    throw new Error(`Unexpected user auth url: ${user.server}`);
+  }
+
+  url.set('pathname', `/~/${realmName}`);
+
+  const config = {
+    schema: schema,
+    sync: {
+      user,
+      url: url.href
+    }
+  };
+
+  const _Realm = user.constructor._realmConstructor;
+
+  return new Promise((resolve, reject) => {
+    _Realm._waitForDownload(config, (error) => {
+
+      // FIXME: I don't understand why, but removing the following setTimeout causes the subsequent
+      // setTimeout call (when resolving the promise) to hang on RN iOS. 
+      // This might be related to our general makeCallback issues.
+      setTimeout(() => {}, 1);
+
+      if (error) {
+        reject(error);
+      }
+      else {
+        try {
+          let syncedRealm = new _Realm(config);
+          user[specialPurposeRealmsKey][realmName] = syncedRealm;
+          //FIXME: RN hangs here. Remove when node's makeCallback alternative is implemented
+          setTimeout(() => resolve(syncedRealm), 1);
+        } catch (e) {
+          reject(e);
+        }
+      }
+    });
+  });
+}
+
+function createInManagementRealm(user, modelName, modelInitializer) {
+  return getSpecialPurposeRealm(user, '__management', managementSchema)
+    .then(managementRealm => {
+      return new Promise((resolve, reject) => {
+        try {
+          let o;
+
+          const listener = () => {
+            if (!o) {
+              return;
+            }
+
+            if (typeof o.statusCode === 'number') {
+              managementRealm.removeListener('change', listener);
+
+              if (o.statusCode === 0) {
+                setTimeout(() => resolve(o), 1);
+              }
+              else {
+                const e = new Error(o.statusMessage);
+                e.statusCode = o.statusCode;
+                e.manamgementObject = o;
+                setTimeout(() => reject(e), 1);
+              }
+            }
+          }
+
+          managementRealm.addListener('change', listener);
```

Comment:
Doesn't js support object level notifications?

Reply:
Not yet: https://github.com/realm/realm-js/issues/763

Reply:
no

Reply:
😿